src/envs/playroom.py

```diff
@@ -0,0 +1,845 @@
+            _terminal=self._Pick_terminal)
+        self._Stack = ParameterizedOption(
+            # variables: [robot, object on which to stack currently-held-object]
+            # params: [delta x, delta y, delta z]
+            "Stack", types=[self._robot_type, self._block_type],
+            params_space=Box(-1, 1, (3,)),
+            _policy=self._Stack_policy,
+            _initiable=self._Stack_initiable,
+            _terminal=self._Stack_terminal)
+        self._PutOnTable = ParameterizedOption(
+            # variables: [robot]
+            # params: [x, y] (normalized coordinates on the table surface)
+            "PutOnTable", types=[self._robot_type],
+            params_space=Box(0, 1, (2,)),
+            _policy=self._PutOnTable_policy,
+            _initiable=self._PutOnTable_initiable,
+            _terminal=self._PutOnTable_terminal)
+        # Objects
+        self._robot = Object("robby", self._robot_type)
+        self._door1 = Object("door1", self._door_type)
+        self._door2 = Object("door2", self._door_type)
+        self._door3 = Object("door3", self._door_type)
+        self._door4 = Object("door4", self._door_type)
+        self._door5 = Object("door5", self._door_type)
+        self._door6 = Object("door6", self._door_type)
+        self._doors = (self._door1, self._door2, self._door3,
+                       self._door4, self._door5, self._door6)
+        self._dial = Object("dial", self._dial_type)
+
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        prev_x = state.get(self._robot, "pose_x")
+        prev_y = state.get(self._robot, "pose_y")
+        x, y, z, rotation, fingers = action.arr
+        cls = PlayroomEnv
+        # Infer which transition function to follow
+        # In boring room next to table
+        if self._NextToTable_holds(state, (self._robot,)):
+            # Interact with blocks
+            if (cls.table_x_lb < x < cls.table_x_ub) \
+                and (cls.table_y_lb < y < cls.table_y_ub) \
+                and self._robot_is_facing_table(action):
+                if fingers < 0.5:
+                    transition_fn = self._transition_pick
+                elif z < self.table_height + CFG.playroom_block_size:
+                    transition_fn = self._transition_putontable
+                else:
+                    transition_fn = self._transition_stack
+            # Move elsewhere in boring room
+            elif (0.0 < x < 30.0) and (0.0 < y < 30.0):
+                transition_fn = self._transition_move
+            # Move to hallway or to playroom
+            elif ((30.0 < x < 110.0 and 10.0 < y < 20.0) \
+                    or (110.0 < x < 140.0 and 0.0 < y < 30.0)) \
+                    and self._robot_can_move(state, action):
+                transition_fn = self._transition_move
+            else:  # Invalid
+                return state
+        # In boring room next to door
```

Comment:
hmm I'm wondering if we can consolidate some of this so that we just have one chunk of code to handle opening/closing any door. maybe it would be helpful to break it up like: first handle the movement part of the action. then, after moving, check if we're next to any door and open or close it. would that work?